File: sdk/typescript/src/resources/signals.ts
import type { BastionHttpClient } from "../http.js";

export interface SignalListOptions { limit?: number; offset?: number; horizon?: string }

export class SignalsResource {
  constructor(private readonly http: BastionHttpClient) {}
  listTop(options: SignalListOptions = {}): Promise<unknown> { return this.http.get("/signals/top", { query: { ...options } }); }
  latest(options: { limit?: number } = {}): Promise<unknown> { return this.http.get("/signals/latest", { query: options }); }
  get(signalId: string | number): Promise<unknown> { return this.http.get(`/signals/${signalId}`); }
  getEvidence(signalId: string | number): Promise<unknown> { return this.http.get(`/signals/${signalId}/evidence`); }
  getDeliveryLogs(signalId: string | number): Promise<unknown> { return this.http.get(`/signals/${signalId}/delivery-logs`); }
  getExplanation(signalId: string | number): Promise<unknown> { return this.http.get(`/signals/${signalId}/explanation`); }
  getRecommendations(signalId: string | number): Promise<unknown> { return this.http.get(`/signals/${signalId}/recommendations`); }
}

File: sdk/python/bitcoin_bastion_sdk/resources/signals.py
from __future__ import annotations

from collections.abc import Iterator
from typing import Any

from bitcoin_bastion_sdk.pagination import iter_paginated
from bitcoin_bastion_sdk.resources.base import AsyncBaseResource, BaseResource


class SignalsResource(BaseResource):
    def list_top(self, *, limit: int = 20, offset: int = 0, raw: bool = False) -> Any:
        return self._get("/signals/top", params={"limit": limit, "offset": offset}, raw=raw)

    def iter_top(self, *, limit: int = 100) -> Iterator[dict[str, Any]]:
        yield from iter_paginated(lambda *, limit, offset: self.list_top(limit=limit, offset=offset), limit=limit)

    def latest(self, *, raw: bool = False) -> Any:
        return self._get("/signals/latest", raw=raw)

    def get(self, signal_id: int | str, *, raw: bool = False) -> Any:
        return self._get(f"/signals/{signal_id}", raw=raw)

    def get_evidence(self, signal_id: int | str, *, raw: bool = False) -> Any:
        return self._get(f"/signals/{signal_id}/evidence", raw=raw)

    def get_delivery_logs(self, signal_id: int | str, *, raw: bool = False) -> Any:
        return self._get(f"/signals/{signal_id}/delivery-logs", raw=raw)

    def recommendations(self, signal_id: int | str, *, raw: bool = False) -> Any:
        return self._get(f"/signals/{signal_id}/recommendations", raw=raw)

    def explanation(self, signal_id: int | str, *, raw: bool = False) -> Any:
        return self._get(f"/signals/{signal_id}/explanation", raw=raw)


class AsyncSignalsResource(AsyncBaseResource):
    async def list_top(self, *, limit: int = 20, offset: int = 0, raw: bool = False) -> Any:
        return await self._get("/signals/top", params={"limit": limit, "offset": offset}, raw=raw)

    async def latest(self, *, raw: bool = False) -> Any:
        return await self._get("/signals/latest", raw=raw)

    async def get(self, signal_id: int | str, *, raw: bool = False) -> Any:
        return await self._get(f"/signals/{signal_id}", raw=raw)

    async def get_evidence(self, signal_id: int | str, *, raw: bool = False) -> Any:
        return await self._get(f"/signals/{signal_id}/evidence", raw=raw)

    async def get_delivery_logs(self, signal_id: int | str, *, raw: bool = False) -> Any:
        return await self._get(f"/signals/{signal_id}/delivery-logs", raw=raw)
